File: mobile/app/lib/utils/wikipedia_image_helper.dart
import 'package:flutter/foundation.dart';

/// Utility methods for converting low-resolution Wikipedia thumbnail URLs
/// into higher quality variants that look sharper in the app.
class WikipediaImageHelper {
  const WikipediaImageHelper._();

  /// Converts a Wikipedia thumbnail URL to its full resolution version.
  ///
  /// Wikipedia thumbnail URLs typically look like:
  /// `.../thumb/a/ab/Filename.jpg/240px-Filename.jpg`.
  /// This method removes the `/thumb/` segment and the trailing `240px-`
  /// suffix so that the original asset is returned.
  static String getHighResolutionUrl(String imageUrl) {
    if (imageUrl.isEmpty) return imageUrl;

    try {
      if (imageUrl.contains('/thumb/') && imageUrl.contains('px-')) {
        var cleanUrl = imageUrl.replaceFirst('/thumb/', '/');
        final lastSlash = cleanUrl.lastIndexOf('/');

        if (lastSlash != -1) {
          final path = cleanUrl.substring(0, lastSlash);
          final filename = cleanUrl.substring(lastSlash + 1);

          final pxIndex = filename.indexOf('px-');
          if (pxIndex != -1) {
            final actualFilename = filename.substring(pxIndex + 3);
            return '$path/$actualFilename';
          }
        }
      }

      if (imageUrl.contains('.org/') && imageUrl.contains('/thumb/')) {
        final uri = Uri.parse(imageUrl);
        if (uri.queryParameters.containsKey('width')) {
          final params = Map<String, String>.from(uri.queryParameters)
            ..remove('width')
            ..remove('height');
          var cleanUrl = uri.replace(queryParameters: params).toString();
          cleanUrl = cleanUrl.replaceFirst('/thumb/', '/');
          if (cleanUrl.contains('px-')) {
            return _removeSizeSuffix(cleanUrl);
          }
          return cleanUrl;
        }
      }
    } catch (err) {
      debugPrint('Wikipedia URL conversion failed: $err');
      return imageUrl;
    }

    return imageUrl;
  }

  /// Returns an optimized image URL sized appropriately for mobile/tablet.
  static String getOptimizedUrl(String imageUrl, {int targetWidth = 1024}) {
    final baseUrl = getHighResolutionUrl(imageUrl);

    if (baseUrl.contains('wikimedia.org/') || baseUrl.contains('wikipedia.org/')) {
      try {
        var workingUri = Uri.parse(baseUrl);
        final segments = List<String>.from(workingUri.pathSegments);

        if (!segments.contains('thumb') && segments.length > 2) {
          segments.insert(2, 'thumb');
          workingUri = workingUri.replace(pathSegments: segments);
        }

        final uriString = workingUri.toString();
        final lastSlash = uriString.lastIndexOf('/');
        if (lastSlash != -1) {
          final filename = workingUri.pathSegments.last;
          final path = uriString.substring(0, lastSlash);
          return '$path/${targetWidth}px-$filename';
        }
      } catch (err) {
        debugPrint('Wikipedia optimize failure: $err');
      }
    }

    return baseUrl;
  }

  static String _removeSizeSuffix(String url) {
    final lastSlash = url.lastIndexOf('/');
    if (lastSlash != -1) {
      final path = url.substring(0, lastSlash);
      var filename = url.substring(lastSlash + 1);
      filename = filename.replaceFirst(RegExp(r'^\d+px-'), '');
      return '$path/$filename';
    }
    return url;
  }

  static bool isWikipediaImage(String url) =>
      url.contains('wikimedia.org/') || url.contains('wikipedia.org/');

  static String getDisplayUrl(String? imageUrl, {int targetWidth = 1024}) {
    if (imageUrl == null || imageUrl.isEmpty) return '';
    if (isWikipediaImage(imageUrl)) {
      return getOptimizedUrl(imageUrl, targetWidth: targetWidth);
    }
    return imageUrl;
  }
}

extension WikipediaImageExtension on String {
  String get highResolution => WikipediaImageHelper.getHighResolutionUrl(this);

  String optimized({int width = 1024}) =>
      WikipediaImageHelper.getOptimizedUrl(this, targetWidth: width);
}

File: frontend/src/utils/wikipediaImageHelper.ts
const normalizeProtocol = (url: string): string =>
  url.startsWith('//') ? `https:${url}` : url;

const stripSizeSuffix = (url: string): string => {
  const lastSlash = url.lastIndexOf('/');
  if (lastSlash === -1) return url;

  const path = url.substring(0, lastSlash);
  const filename = url.substring(lastSlash + 1).replace(/^\d+px-/, '');
  return `${path}/${filename}`;
};

export const isWikipediaImage = (url?: string | null): boolean => {
  if (!url) return false;
  const normalized = normalizeProtocol(url);
  return normalized.includes('wikimedia.org/') || normalized.includes('wikipedia.org/');
};

export const getHighResolutionUrl = (imageUrl: string): string => {
  if (!imageUrl) return imageUrl;

  const normalizedUrl = normalizeProtocol(imageUrl);

  try {
    if (normalizedUrl.includes('/thumb/') && normalizedUrl.includes('px-')) {
      let cleanUrl = normalizedUrl.replace('/thumb/', '/');
      const lastSlash = cleanUrl.lastIndexOf('/');
      if (lastSlash !== -1) {
        const path = cleanUrl.substring(0, lastSlash);
        const filename = cleanUrl.substring(lastSlash + 1);
        const pxIndex = filename.indexOf('px-');

        if (pxIndex !== -1) {
          const actualFilename = filename.substring(pxIndex + 3);
          return `${path}/${actualFilename}`;
        }
      }
    }

    if (normalizedUrl.includes('.org/') && normalizedUrl.includes('/thumb/')) {
      const uri = new URL(normalizedUrl);
      if (uri.searchParams.has('width')) {
        const params = new URLSearchParams(uri.searchParams);
        params.delete('width');
        params.delete('height');
        uri.search = params.toString() ? `?${params}` : '';
        let cleanUrl = uri.toString().replace('/thumb/', '/');
        if (cleanUrl.includes('px-')) return stripSizeSuffix(cleanUrl);
        return cleanUrl;
      }
    }
  } catch (error) {
    console.warn('Wikipedia URL conversion failed:', error);
  }

  return normalizedUrl;
};

export const getOptimizedUrl = (imageUrl: string, targetWidth = 1024): string => {
  const baseUrl = getHighResolutionUrl(imageUrl);
  if (!isWikipediaImage(baseUrl)) return baseUrl;

  const sanitizedWidth = Math.max(1, Math.round(targetWidth));

  try {
    const url = new URL(baseUrl);
    const segments = url.pathname.split('/').filter(Boolean);

    if (!segments.includes('thumb') && segments.length > 2) {
      segments.splice(2, 0, 'thumb');
    }

    if (!segments.length) return baseUrl;

    const filename = segments[segments.length - 1];
    const prefixSegments = segments.slice(0, -1);
    const prefix = prefixSegments.length
      ? `${url.origin}/${prefixSegments.join('/')}`
      : url.origin;

    return `${prefix}/${sanitizedWidth}px-${filename}`;
  } catch (error) {
    console.warn('Wikipedia optimize failure:', error);
    return baseUrl;
  }
};

export const getDisplayUrl = (imageUrl?: string | null, targetWidth = 1024): string => {
  if (!imageUrl) return '';
  if (isWikipediaImage(imageUrl)) {
    return getOptimizedUrl(imageUrl, targetWidth);
  }
  return normalizeProtocol(imageUrl);
};

export default {
  getHighResolutionUrl,
  getOptimizedUrl,
  getDisplayUrl,
  isWikipediaImage,
};
